### ISSUE TITLE
Samsung Galaxy S7 Edge (Android) - Streaming not optimal

### ISSUE THREAD

**Issue by XaivierX** [moonlight-stream/moonlight-android#349]
Hi there,

First off, thanks for creating this project.  Next, if there is anything I can do to assist or troubleshoot on my end (logs, settings, etc), let me know, I'll be happy to do everything I can to get you as much info as possible.

Issue:  On Samsung Galaxy S7 Edge (Android Marshmallow), the stream quality is jumpy, jerky, and very poor compared to streaming on Windows (on a late 2000's laptop) or OSX.  I have not tested yet on an iPhone, but I can, if desired. In addition, it is also comparable, if not worse, to the quality of streaming that I encountered on my Galaxy Note 3 before upgrading to this device.

I can say for sure it has nothing to do with the target mbps, the wireless router, or the host pc.  Multiple tests on other streaming devices have proved that the quality can be, and is, top-notch and smooth with no issues.  It does not matter the resolution selected for streaming (720, 1080, 30 or 60fps).  My decoding latency is only around 15ms on the device as well.

If I had to describe in detail the issue, the stream seems to display perfectly well, especially when not much on the screen is happening, but once the screen moves, such as a 2d scroller, the streaming shows signs of micro-studdering and has multiple instances of bigger frame jumps (such as the character is in one place on the screen during a jump, then, instead of taking the entire process smoothly, the scene may hang for a quarter second, then update with the character's current position, and then continue on as smooth as it was before the hang.

I've turned off extra services such as location service, mobile data, sync, etc on the device with no apparent improvement of streaming quality.  So, with all of my resources exhausted, I'm posting this issue here in hopes that perhaps we can discover the cause, and hopefully have it addressed.

**Comment by JasonS7notG5** [comment]
I also have the S7 edge (G935F Exynos SoC / Mali GPU) with Nougat and I can confirm that gameplay is not smooth 60fps, it has micro stutters (frame pacing?). Decoding is less than 15ms all the time (720p-1080p 60fps 8-10Mbit H.264). I tested this with Call of Duty: Modern Warfare Remastered where it gets really obvious that there is a problem with how Moonlight utilizes the SoC, I mean the S7 edge is not a cheap phone and uses top hardware. Why am I sure that this must be a hardware/software related problem specific to the S7 edge? My ASUS tablet which cost me like $99 with an Intel Atom Z3745 SoC on the same network with the same WiFi signal strengh even on 2.4GHz can play smooth 60fps with no micro stutters. I tested both 5GHz and 2.4GHz on my S7 edge and it didn't matter on which network I was connected. I would say that anyone with an S7 / S7 edge atleast with the Exynos SoC will have the same experience like me and XaivierX. This is not network, firewall, config, game etc. related but how Moonlight handles this with the Exynos SoC. Please fix this, I really dont want to miss out playing PC games on the go with smooth framerates on a beautiful AMOLED display. Thanks.

**Comment by rob151515** [comment]
I have the same phone S7 edge (G935F Exynos SoC / Mali GPU) only with Marshmallow and can confirm the same behavior. I have tried everything I can think of as well as all suggestions I have found on the internet (different display drivers etc) and none have resolved the issue. Have to agree with both XaivierX and JasonS7notG5 that this appears to be an issue with Moonlight and not my PC/Network. I have tested with an old Samsung 10" tablet and the streaming is smooth as butter. The hardware involved in my setup is a i7 Windows 10 PC with GTX 780ti etherneted to my r7000 router then 5ghz wifi from there to my S7. When gaming over Moonlight I am finding that the S7 is smooth for about 2-3 seconds then stutters then is smooth again for 2-3 seconds then stutters and so on... I have tried changing resolution/bitrate but nothing seems to make a difference. Hopefully someone can get to the bottom of this. I am happy to provide any diagnostics or do any testing that can help with resolving this problem.

**Comment by JasonS7notG5** [comment]
I made screen recordings from the S7 edge while playing Call of Duty: Modern Warfare Remastered and while watching Mass Effect: Andromeda 60fps gameplay on YouTube. Here are the links:

https://www.youtube.com/watch?v=OdoY2oIQRzA
https://www.youtube.com/watch?v=zfj9MgYhroY

Moonlight settings: 720p 60fps 8Mbit H.264
Latency was approx. 10/8 ms.
Recorded with Samsung Game Tools.

Original video to download (without YouTube compression etc.):
http://www.mediafire.com/file/fhkzy3digo8o1j4/Moonlight+-+Call+of+Duty+Modern+Warfare+Remastered.mp4
http://www.mediafire.com/file/7srddfw9l9l9e5n/Moonlight+-+Mass+Effect+Andromeda+Gameplay.mp4

I would have liked to record my ASUS tablet or even my desktop PC but it wasnt possible because there is no screen recorder for Android 4.4 non root and somehow Shadowplay didn't worked while streaming. I am using the newest GeForce Experience 3.3.0.100.

On my ASUS tablet it plays the game and the YouTube video smooth with occasionally lag because of 2.4GHz which I have noticed but overall there are no micro stutters all the time like on my S7 edge.

**Comment by JasonS7notG5** [comment]
Well after hours of trial and error I can absolutely say the micro stutters are because of the hardware or how Moonlight utilizes the hardware. I used Moonlight with USB tethering and GameBench to show the FPS on my S7 edge and compared it to the FPS counter on my monitor. While my PC / monitor shows 60FPS and feels like it, my S7 edge shows drops every second from 60 to 58 to 59 to 52 to 5x etc. = micro stutters. This does not only look not smooth but also affects the gameplay, for example when playing CoD: MW I get better results when watching the monitor while I play worse when watching the S7 edge (I played with mouse and keyboard attached to the PC).

I made some videos showing this micro stutter problem:
**[This first one](https://www.youtube.com/watch?v=jsOXC6fW3qU)** shows me playing CoD: MW with Moonlight using USB tethering and GameBench to show the FPS recorded with internal Samsung Game Tools app. You can clearly see how the FPS jumps all over the place within 50 to 60 FPS which makes the gameplay feel not laggy but also not smooth. [_The first and third run in CoD: MW training mission where I got better results I watched the monitor while playing, the second and fourth run with worse results on the S7 edge_]
[The second one](https://www.youtube.com/watch?v=YFzQ0SIKisE) also shows me playing CoD: MW with the same config as the first one but this time recorded with an iPhone 4S (30fps recording unfortunately) to show there is no influence in the recording software (Samsung Game Tools). Still micro stutters here.
And [the last one](https://www.youtube.com/watch?v=r2YNXwNoGKQ) shows the FPS counter on the S7 edge compared to the FPS counter on the monitor.

Here are screenshots showing the micro stutters:
Related to the first video:
-[GameBench Session Log screenshot](http://imgur.com/a/1wOzT)
-[GameBench framerate per second pictures](https://mega.nz/#F!991UAYoI!CwEhDh3szzxpuz_LYQdBZA)
Related to the second video:
-[GameBench Session Log screenshot](http://imgur.com/a/pt3MR)
-[GameBench framerate per second pictures](https://mega.nz/#F!w4FCXLZI!R00ebcEeWo2d5-8UqDGgzw)

I hope this micro stutters problem can get fixed very soon. There is no other streaming app getting even close to Moonlight, I tried most of them like Remotr and others but they are full of ads and perform really bad, not playable. I am thankful for Moonlight and hope it will get developed further. Thanks.

**Comment by XaivierX** [comment]
Can anyone confirm that turning off adaptive display (under Settings > Display > Screen Mode) helps at all?  Not sure I'm I'm actually seeing a difference or if I'm being hopeful.

**Comment by JasonS7notG5** [comment]
No there is no difference. Just tested it with USB tethering in offline mode with AMOLED cinema, picture, basic and dropping the resolution to FHD (Nougat). Still micro stutters...

**Comment by JasonS7notG5** [comment]
Summary/Conclusion from me regarding Moonlight on the S7 edge and other devices:

I am using Moonlight 4.8.3 (Android) / 1.1.4 (iOS) with GeForce Experience 3.3.0.100.
I have tested Moonlight on different devices:
***Samsung Galaxy S7 edge** (Exynos 8890 variant, Android 7.0 Nougat)
***ASUS Memo Pad 8** (Intel Atom Z3745, Android 4.4.2 KitKat)
***Apple iPhone 4S** (Apple A5, iOS 9.3.5)

For both Android devices I used [GameBench](http://www.guidingtech.com/60101/record-exact-frame-rate-android-games/) to display and capture the FPS, on iOS this was not possible (without jailbreaking). While the ASUS tablet and the iPhone 4S connected with USB tethering on my PC had constant smooth streaming playback, the S7 edge did not. The ASUS tablet and the iPhone 4S while working had some milliseconds of video delay but definitly playable and enjoyable. The S7 edge has micro stutters, every second (or less) it switches from 60FPS to 50-something FPS back to 60FPS etc. resulting in a choppy streaming playback which means no smooth gameplay possible. This has nothing to do with network since I used USB tethering while being in offline mode. With the iPhone 4S on WiFi 2.4GHz it was even possible to have a very good streaming playback (without knowing how much FPS it did have), no micro stutters but sometimes a little lag which isn't that much interrupting the gameplay. The ASUS tablet also on WiFi 2.4GHz can get - when the cheap WiFi antenna inside it works... - the same results like the iPhone 4S. Now something positive for the S7 edge: It may not have the same smooth streaming playback versus the ASUS tablet and iPhone 4S but it has the best response time for input/video. On the S7 edge it looks and plays like if there is no video delay like the other both devices have. How can it have better response time or no delay for the video stream and input when the video itself is full of micro stutters? Maybe it has better response times because of the micro stutters or how Moonlight handles the S7 edge differently, I don't know. But I personally would say it first needs smooth streaming playback and right next to it keep the response times it has right now. It makes no fun playing a game full with micro stutters because it doesnt feel like real 60fps.

Here are some data for you guys, please check them out and look what I mean:
[S7 edge showing micro stutters (captured internally with Samsung Game Tools)](https://www.youtube.com/watch?v=jsOXC6fW3qU)
(look at the description for GameBench FPS logs)

[S7 edge showing micro stutters (recorded externally with an iPhone 4S)](https://www.youtube.com/watch?v=YFzQ0SIKisE)
(look at the description for GameBench FPS logs)

[Comparing the FPS counter on the S7 edge with the FPS counter on the PC / monitor](https://www.youtube.com/watch?v=r2YNXwNoGKQ)

[ASUS tablet versus PC FPS counter and delay comparison in slow motion in Call of Duty](https://www.youtube.com/watch?v=L7i76_PdaRo)

[ASUS tablet versus PC FPS counter and delay comparison in slow motion in YouTube video](https://www.youtube.com/watch?v=pxe9-lhvsow)

[iPhone 4S versus PC FPS counter and delay comparison in slow motion in Call of Duty](https://www.youtube.com/watch?v=jIplFGcyB1I)

[iPhone 4S versus PC FPS counter and delay comparison in slow motion in YouTube video](https://www.youtube.com/watch?v=CzP95FV6Rpc)

(_I have some GameBench log files for the ASUS tablet and iPhone 4S but don't think they are necessary, I looked at them and they are showing constant 60-ish FPS (60 most of the time, going to 59 and 61 sometimes, not considering this as micro stutters). And I also have 240fps videos (slow motion) from ASUS tablet and iPhone 4S WiFi 2.4GHz gameplay._)


I would be happy if anyone (doesn't matter which device) could also use GameBench http://www.guidingtech.com/60101/record-exact-frame-rate-android-games/ and post their results so we can make comparisons and see which device needs better support from Moonlight. I mean the S7 edge is one of the best smartphones right now, top hardware from 2016 and it loses against an iPhone from 2011 and an old cheap ASUS tablet? That doesn't make any sense.

**Comment by OakyProductions** [comment]
Hi guys!

I have tried to stream to my Samsung Galaxy S7 Edge, and a different problem is happening to me: every eight seconds, the image lags a little, and then it comes back to normal quality, but, as I said, it happens every eight seconds with no exception, and it makes games unplayable.

I think that this is happening because Android 7.0, because when I was on 6.0, I never had any problem with streaming via Moonlight.

Do you think this can be solved? It should be solved by Moonlight or by Samsung itself updating the software of the phone? I said that because PS4's Remote Play doesn't work anymore on my phone, and in 6.0 it did.

This is a very confusing thing, I hope it will be solved soon.

**Comment by JasonS7notG5** [comment]
If PS4 Remote Play also not work then I would say it is Android 7.0 or your WiFi but less the hardware and your problem is a seperate one, has nothing to do with our problems which is a Moonlight problem.

Plug in your S7 edge on your PC and then go to your phone settings and enable 'USB tethering'. After that your PC will connect to a network created by your S7 edge and then go into offline mode (no WiFi no 4G) and start Moonlight. Play a game and I bet your problem is fixed.

**Comment by OakyProductions** [comment]
No, I didn't flash anything. I see a relation between my problem with Moonlight and with Remote Play, but my main focus is on Moonlight (which works without any problems on my iPad Air).

On the other hand, I tried what you said and my computer doens't connect to my phone, but I don't understand why this would solve the lag every eight seconds. Anyways, thanks for the help!

**Comment by JasonS7notG5** [comment]
Why does no one from the dev team response to this? THIS IS A HUGE PROBLEM, this needs to be fixed. But it seems like they dont care anymore for Moonlight. What I found out is that many newer devices (if not all) with Android 6.0+ has this micro stutter problem. The devs think because only some notice and report this this must be only a problem for few people which may have problems with their setup. But thats wrong. I proved it here with the videos I uploaded on YouTube. I proved it with GameBench (FPS counter). Atleast, really atleast, the Galaxy S7 Exynos variant doesnt matter on which Android version has micro stutters. But if you look at others with newer devices and Android 6.0+ they also say they have micro stutters, look here:

[Microstutter Problem - Testing and Results](https://github.com/moonlight-stream/moonlight-android/issues/214)
[Nothing but stuttering on Android. Mostly perfectly smooth on iOS.](https://github.com/moonlight-stream/moonlight-android/issues/255)
[micro stuttering Moonlight Game Streaming](https://github.com/moonlight-stream/moonlight-android/issues/326)
[Microstuttering with "poor network connection" on Android](https://github.com/moonlight-stream/moonlight-android/issues/328)

And if anyone wants to proof that they dont have any micro stutters on newer devices, please use GameBench and post your logs here (GameBench also makes screenshots every second including FPS counter).

**Comment by colin-foster-in-advantage** [comment]
What I have noticed on a similar application is the VideoReceive thread blocks for too long, specifically on the system call to the network driver. I couldn't reliably recreate it in other applications, but my suspicion was it had to do with the garbage collector. It might be worth trying to verify whether that is happening by recording a method trace when you're seeing the jitter. 

Another theory is that I-frames are being sent out every 8 seconds, which causes inconsistency. I'm not sure if you could get that from the logs / traces or if you'd need a wireshark capture, which adds another level of analysis.

**Comment by redlight5** [comment]
I have same issue with my Galaxy S7 and Nexus 5X (I did not test 5X that much). Works perfectly on my Shield Tablet.

**Galaxy S7** - Missing frames every second or two causing skips in the image. To me it feels like Moonlight is not in sync or that the decoding of the video stream is not up to par on this device, I also get quite a bit of audio delay even the video delay is very low. PC wired to Nethawk R6700 and phone connected through AC wifi, good signal and using TCP to capture the bitrate between phone and PC gave me results of around 230mbps.

**Shield Tablet** - Works perfectly in both official Shield Hub and Moonlight, no single dropped frame or judder. Connected to same AC and this device is known to have quite bad wifi chip.

Now this brings up my question: Does Moonlight get the stream from the PC in different encoding using the official Shield tablet compared to S7? I've read that Nvidia restricts some capturing format or codec based from third-party apps. But would that also affect Moonlight on this device also?

Anyways, I really hope we find some solution to this, I'm completely sold to using Stream Theater with Gear VR and the stuttering is the only thing that brings down the experience. Going to keep following!

**Comment by JasonS7notG5** [comment]
I also have noticed the audio delay on my S7 edge while on my iPhone 4S it just works perfectly fine. But what I noticed on my S7 edge is how much better the response time (input lag) is versus my other (old) devices. This is a huge plus and needs to stay. I really hope this micro stutters problem will get fixed soon. One of the devs also has a Galaxy S7...

**Comment by redlight5** [comment]
I'm probably going to take another look over the weekend myself as well if I get the time. Another thing to point out is my S7 is the Qualcomm 820 model.

**Comment by JasonS7notG5** [comment]
I have the Exynos model. When I use USB tethering there is no (or minimal) audio delay while when using it on WiFi (5GHz, 2.4GHz) there is noticeable audio delay. On my iPhone 4S there is no audio delay (USB tethering and also WiFi 2.4GHz).

**Comment by JasonS7notG5** [comment]
This is 100% Android 6.0 and newer related! I just tested this with my Razer Forge TV which I got today.

> So I received my Razer Forge TV. It had Android 5.1.1 on it. I installed Moonlight and started playing Call of Duty: Modern Warfare Remastered. It had no micro stutters on WiFi 5GHz. But the response time from the controller / keyboard+mouse was not instant, it had a litte bit of input lag. Then unfortunately it auto-updated to Android 6.0 while I was playing CoD with Moonlight..... so I couldn't make any GameBench tests to prove that their was no micro stutters but input lag. Now its on Android 6.0.1. What I noticed immediately is that the response times (input lag) is so much better BUT now it has micro stutters again. And audio delay. Somehow exactly the same as my S7 edge (Exynos) on Android 7.0 Nougat. THIS IS 100% ANDROID 6.0 MARSHMALLOW AND NEWER RELATED, CONFIRMED! Every single Android 6.0+ device has micro stutters! Nothing network, latency or something else related. The only thing which may be network related is the audio delay BUT on Android 5.1.1 there was no audio delay. Why it may be network related? Because when I used USB tethering on my S7 edge (this means no WiFi, instant perfect connection between PC and device with USB) there was no audio delay but micro stutters where still visible. So please devs FIX THIS ISSUE.

I will try to use GameBench (if it is possible on Android TV) on the Razer Forge TV with tethering. But if you look at others who also wrote they have problems on devices with Android 6.0 and newer... this must be the problem.

https://github.com/moonlight-stream/moonlight-android/issues/214
https://github.com/moonlight-stream/moonlight-android/issues/255
https://github.com/moonlight-stream/moonlight-android/issues/326
https://github.com/moonlight-stream/moonlight-android/issues/328

**Comment by Chuacheneguer** [comment]
In my samsung s4 with rom lollipop works perfect but with roms marshmallow and nougat i have stuttering

**Comment by JasonS7notG5** [comment]
Thats what I said to the dev of Moonlight but he said that 70.000+ people are on Android 6.0 and higher and if there was a problem, many more would complain. He doesn't really believe me... I said him that Android 4.4 KitKat and 5.1 Lollipop have smooth video playback but input lag while Android 6.0 Marshmallow and 7.0 Nougat have no input lag but micro stutters in video playback (not smooth).

I can only guess why the other 70.000 did not report this problem. First, micro stutters are not a game breaker for many people, I would say most dont even notice it in the first place but not because there is no micro stutters but because they dont know much about smooth gameplay (consistent 60FPS = 16.7ms frame times). Also many app users are just enjoying the app and they dont know how or really want to report problems, since like I said this is not a 'unplayable' problem.

If the devs really want to prove that I am wrong and others too and we are a minority which screwed up the setup which is why we have problems, then **please Moonlight devs, use GameBench on your Android 6.0+ devices** and post your results here. Here have again a look at my GameBench graph and screenshots with FPS counter.

http://imgur.com/a/pt3MR
https://mega.nz/#F!w4FCXLZI!R00ebcEeWo2d5-8UqDGgzw

**Comment by Trasen** [comment]
@JasonS7notG5 :  Not sure it's a 6.0+ issue, however the app does suffer from microstutters. I solved my micro-stutters by reinstalling the Nvidia drivers by first running [Driver Sweeper](http://www.guru3d.com/content-page/guru3d-driver-sweeper.html) and then just installing the newest released Nvidia driver.

Could you please try this and post the results? I am courious since it was the only thing i had to do to get rid of that pesky issue.

Best regards.

**Comment by cgutman** [comment]
I never said there wasn't a bug. I've seen it with my own eyes. I've also spent many hours successfully reproducing the bug on my own devices.

My point is that I work on this project as a hobby in my limited free time. If it's not breaking almost everyone (like a new GFE update totally breaking streaming), it's hard to be motivated after working a 9 hour day at a real job, to sit down and work for another 6 hours on this issue.

As I said in email, I will continue working on it but progress is slow.

**Comment by XaivierX** [comment]
@Trasen : I was really skeptical about your solution, but I went ahead and cleaned my drivers and updated with the latest NVidia driver, and the streaming is much better.  Thank you!

Makes me believe, if it works for a couple others, that it would point to the streaming service + latest android version as the cause itself.  Perhaps NVidia finally addressed the issue(?)

**Comment by XaivierX** [comment]
@cgutman : Thanks for your dev on the app, we all appreciate your hard work.

**Comment by Trasen** [comment]
@XaivierX : No problem bud! However, this might not even be an issue with the streaming service at all.

In general stack-updating Nvidia drivers gives performance issues, there's really no exception. Hence the reason why you should always do a driver sweep before any major Nvidia release.

If it's caused by Geforce Experience or the actual driver though, that's a question that would take re-creating the problem fairly consistently, which i've yet to be able to do.

But i'm really happy it worked for you :) Let's see if anyone else confirms that this is a valid solution.

Best regards!

**Comment by Chuacheneguer** [comment]
My experience with moonlight is:
Samsung S4: ROMS lollipop: works perfect, official or customs ROMS.
                      ROMS marshmallow and nougat:  i have stuttering

motorola Moto G 4g : Lollipop: works perfect

oneplus3t: marshmallow and nougat:  i have stuttering

Xiaomi Mi Max: marshmallow: i have stuttering

My PC: Intel 2500k 8GB RAM windows 10 pro 64bits . nvidia 780ti

**Comment by JasonS7notG5** [comment]
I used Driver Sweeper (now Display Driver Uninstaller), reinstalled newest NVIDIA drivers and GeForce Experience, made a restart and still micro stutters. Tested while USB tethering and also while in WiFi 5GHz / 2.4GHz, it's the same old problem. And I thought the "new" Moonlight 5.0.0 (5.0.1) fixed it but it didn't...

**Comment by cgutman** [comment]
@JasonS7notG5 and anyone else having this issue, please try this build: https://drive.google.com/open?id=0B_5O1xNt8Q9laGFIeG1iMEFhdUk

**Comment by genmatrix** [comment]
Tested out with Pixel C (X1 SOC) and noticed it increased latency readings by about 12MS to a total of 22MS average. Normally sit at 10ish MS at 1080p60.

It runs relativity smooth but I am running a custom kernel in 7.1.2 that showed serious improvements in how it handled packets from moonlight (and in general).

Just thought I'd share. :+1:

**Comment by cgutman** [comment]
@genmatrix Thanks for testing.

The cause of the stutter (assuming this fixes it) was a subtle change that was hardly documented in how presentation times are used in Lollipop vs Marshmallow. In Lollipop, the frame was presented without regard for the presentation time at all, unless you used the special function that let you pass a desired presentation time in, which Moonlight did not use. In Marshmallow, the presentation time is pulled through the pipeline from another API where Moonlight was just passing the current time as the desired presentation time (which seemed sensible at the time). Even though Moonlight's usage of the decoding APIs didn't change, the behavior did because of how the presentation times were now being attached.

Without a presentation time, the frames would be rendered at the next v-sync and if 2 frames were pending on the next v-sync, the 1st would be displayed at the next v-sync and the 2nd displayed on the next v-sync after that.

With a presentation time, the behavior differs. If there are 2 frames with presentation times that land them in the same v-sync window, _all but the last frame are dropped by Android's renderer_. While certainly good from a latency perspective (since extra frames shouldn't pile up in the rendering queue), it causes this weird stutter if a frame is slightly delayed and the next frame comes in earlier than 16 ms from the previous one (due to the previous frame's network delay). I think it's also likely that the actual latency didn't change, but the measurement is now showing a higher number because the dropped frames were falsely lowering the reported decode latency.

The APK I attached returns to the old behavior of always queuing the frame for display, rather than dropping frames that come within the same 16 ms window. I suspect I can do better though by providing a more accurate presentation time using some other APIs. That part I'm still looking into.

**Comment by genmatrix** [comment]
Very interesting and good read. I will be wiping my Pixel C back to stock latest release and see how it behaves as well. Since I don't know how much the custom kernel is influenceing performance, I'd like to help rule that out.

**Comment by chicozeb** [comment]
Hi.

First of all, thank Cameron for your hard work. It´s much appreciated.

I´m using Shield TV. With your latest build I was getting stuttering for the entire session(not that i´ve played that much). With this build that you provided, the stuttering is gone. Well, not gone, but only small stuttering from time to time(2 or 3 seconds of stuttering), that may be related to the GPU being maxed out.

I was having stuttering with your latest builds, but it was like 5 or 6 minutes of perfect stream, then stuttering. Restarting the stream solved it. But it woukld start again. This issue seemed to appear after i updated the shield tv to android 7. I don´t recall having it with android 6(and after nvidia fix the gfe stuttering), but i don´t use moonlight too much.

Thanks.

**Comment by JasonS7notG5** [comment]
@cgutman installed the APK and... it is so much better! It feels consistent and smooth! The micro stutters are gone! :D

But GameBench doesnt work and I feel like it is not "60fps smooth" but "near 60fps smooth" and I only played for like 10 minutes (CoD:MW training level). What I mean is there may be no micro stutters anymore but there is like really minimal input lag which makes it not "exactly 60fps smooth". But I played with mouse and keyboard (_and USB tethering since my 5GHz connections only have like 60% strength_), I think with a controller this would be not an issue. But without these micro stutters it is so much more enjoyable, in fact now I will use Moonlight more often! Thanks for this fix! And sorry if I have been rude to you or anybody else. This was not my intention. I felt like "left behind" after spending so much time and "proving" that there are micro stutters and getting no help plus the disbelief that this could be an issue for more than just me.

(_this could be like a placebo effect since I can not see GameBench results but I dont think it is... it really feels smooth_ :P)

**Comment by daedalus87m** [comment]
I have installed the new APK and unfortunately it's still stuttering for me.
I have to add that I'm using 2 routers, and one is in Bridge mode. When connected to the 1st router everything is (and also was before the new APK) very smooth, but I get heavy stutters when connected to my 2nd router (which is the one reaching living&bedroom).

**Comment by thw0rted** [comment]
Is your PC on a wired connection?  That should help, but you could also look at getting a powerline adapter instead of using a second router in bridge mode -- some introduce more delay than others but at least your end to end quality is likely to be better.  (Maybe you know somebody who you could borrow a pair from, to see if it helps enough to justify a purchase?)

**Comment by Chuacheneguer** [comment]
[quote] 
@JasonS7notG5 and anyone else having this issue, please try this build: https://drive.google.com/open?id=0B_5O1xNt8Q9laGFIeG1iMEFhdUk
[/quote]
This version tested in a Xiaomi Mi Max (android 6) works without stuttering but with much delay

**Comment by daedalus87m** [comment]
@thw0rted My connection looks like this:

Router 1 (in office) gets internet. This router is connected to a hub. My PC is also connected on this hub, and a long cable that is connected to this hub goes through some walls into my house (next to office), where I have another Router (the one I want to connect from my phone).

Unfortunately I don't have a powerline adapter or know anyone who has one.

I get a non stuttering connection when connected to Router 1 (40ms delay, but I want to play turn base games anyway), but on Router 2 it stutters all the time and so does the audio, which is veeeery annoying and basically unplayable.

**Comment by thw0rted** [comment]
Oh! When you said "bridge mode" I assumed it meant that your second router was acting as a wireless repeater.  The powerline adapter was supposed to let you connect your routers to each other directly, but you're saying they're already configured with a direct cable connection.  In that event, all I can suggest is that one or both of your routers might be performing poorly, but you probably already knew that.  Sorry!

**Comment by daedalus87m** [comment]
@thw0rted If the 2nd Router is just not up to the task (best case scenario), getting a new router wouldn't be a problem. I was rather afraid that there could be another problem, like it being a technical limitation making it impossible to properly  reroute it to another router or something ( I'm no specialist :D ).

**Comment by chicozeb** [comment]
I´ve bought an android box (mini m8s II). I´ve put android tv inside(superceleron rom) that reduces the latency on moonlight.

When i tried to stream, i could only get 25 Mbps of maximum bitrate. Anything beyond that  makes the image freeze, while sound continues. Pc network card at 1Gbps.

Now, if i set the pc network card at 100 Mbps full duplex, the problem is solved. I can go all the way up to 100Mbps bitrate.

Everything ethernet connected.

Hope this heps someone.

**Comment by redlight5** [comment]
Major improvement in stuttering department, only small warps here and there but looks 10 times better. But I can also confirm there is significant latency (app reports 16ms for me but it feels more like 60ms). Anyways, happy to see improvement here and hoping there is something to do for the latency (it's too bad to be playable for me).

**Comment by chicozeb** [comment]
After trying on the amlogic s905x box and on shield tv, i don't see any latency differences. 40ms hardware decoding latency on s905x and 1 ms for shield tv. 45Mbps bitrate. Ethernet connected.

**Comment by chicozeb** [comment]
Strange... I'm having less latency using 4k60 and 80Mbps, about 25ms, than using 1080p60 at 40 Mbps, about 40ms, and much less than 720p60 at 15Mbps, about 80ms latency...

Strange... maybe the 4k decoder is better on the s905x...

**Comment by chicozeb** [comment]
Oh, I'm using h.265.

**Comment by JasonS7notG5** [comment]
The latency is the roughly the same for me as before (9-12ms on my S7 edge Exynos variant with Nougat, H.264 1080p 8Mbps and 20Mbps tested) with the provided Moonlight APK from cgutman. WiFi 5GHz both on PC and S7 edge with only 60% signal strength and it works most of the time, no micro stutters anymore.

**Comment by jaysharp29** [comment]
Greatly improved on a moto g4 running marshmallow.  No stuttering using h265 and 1080p 60fps.  

Lingering issues are higher latency (24ms) and frame skip.  The frame skips look like vsync off, but are much preferred to the jitter.

**Comment by redlight5** [comment]
Made a small video to demostrate. First half is the official build which has stuttering (visible in the video) but great low latency. Second half is the unofficial build linked here without stuttering but bad latency. The reported latencies in this test were: Official 13ms, Unofficial 17ms (it's definitely more).

https://drive.google.com/open?id=0B4FaEFFulZWINWZsRW9tYWNaZzA

**Comment by chicozeb** [comment]
The new update from the play store brings back the stuttering for me. Sometimes it does not stutter. The latency is awesome. But it's not worth playing like that. 

The reported latency is the same because is only the video decoder latency, and i think that it doesn't report any other sources of latency.

**Comment by chicozeb** [comment]
After a bit more testing, i am getting more or less the same results that i had with the builds before the last one(not the latest, from 8/6/2017). On shield tv, it's smooth for a few minutes then it starts stuttering. The build before the latest(8/6/2017) gave me the worst performance, with the game always stuttering. I'm testing most of the time with the witcher 3.

The build provided here gives me more latency than the latest but no stutters.

**Comment by JasonS7notG5** [comment]
I think the build on the Play Store does not contain the fix for the micro stutters. The build which you can download here is based on v5.0.1. If you read what has been fixed in v5.0.2 you may find what causes the 'high latency' (17ms is not high).

I think if you lower the bitrates to like 20 Mbps it may solve your problem.

> v5.0.2
> - **Reduced video stream latency**
> - **Fixed video stream glitching at high bitrates**
> - **Fixed compatibility with some HEVC decoders**
> - Fixed PCs not being found automatically on some devices
> - Fixed keyboard d-pads not working (again)
> - Fixed some reported crashes and hangs

Like I said I used the micro stutter fixed 5.0.1 build from cgutman which he posted here. My device is the **S7 edge**, Exynos variant, **Android 7.0**. My graphics card is the **GTX 660**, I can only stream in **H.264**, I have the **newest NVIDIA driver and GeForce Experience** and I am on Windows 10. Moonlight settings are default, **1080p 60fps 20 Mbps**. PC and S7 edge **connected via 5GHz WiFi** to my router, 60% signal strength.

So please if you want to report a problem post your complete setup like I did.

**Comment by JasonS7notG5** [comment]
There is definitely more input lag in the micro stutter fixed version versus the official builds. I dont know if it is 'latency' since I only get 9-12ms, same as the official builds. It feels more like input lag, not that huge bad but noticeable but nowhere as bad as the previous micro stutters situation. This must be in relation with the 'solution' for getting rid of these micro stutters.

Can we have a v5.0.2 based build with the micro stutters fix (per APK)?

**Comment by MichaelMKenny** [comment]
@JasonS7notG5 I've been trying to sum up the courage to ask this also. I really hope Cameron might give us a build with less input lag :) Fingers crossed!

_Sent from my Huawei Nexus 6P using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by MichaelMKenny** [comment]
I've been the APK uploaded in this issue for the past while and performance is awesome! I can't go back to micro stutters on the Play Store build. I just really wish there was latency. 

I don't think it's input lag, as such, as I still have latency when connecting my controller directly to my PC. It just has more of a delay playing back the stream on the device, as apposed to what's been displayed on my PC's monitor.

If Cameron posts a build with fixed latency (5.0.2) I will be so happy and will be using Moonlight for Android all the time!

**Comment by cgutman** [comment]
Here's a new build with the latest changes and the microstuttering fix: https://drive.google.com/open?id=0B_5O1xNt8Q9lVkcxaXpkWTZzTDQ

**Comment by chicozeb** [comment]
Hi Cameron. Again, thanks for your hard work. I´ll test it later.

**Comment by chicozeb** [comment]
Here´s my specs.

Intel i7 3770K 3.9GHz Hyperthreading On
Gigabyte GTX 1070 G1 Gaming (default mode - i don´t use gigabyte software)
16 GB Gskill DDR3 1866MHz cas9
64GB Sandisk SSD (OS) + 480 OCZ Trion 150 (games and apps and steam) + 1TB Western Digital Blue + 1TB Samsung HDD)
Asus P8Z77V-LK(realtek ethernet card)
650Watt Antec Power Supply625782
Lg E2350 1080p 60Hz monitor
Huawei router HG8247H (ISP provided)
Windows 10 Pro Founders Edition
Latest drivers and GFE.

I use ethernet connected all the time. No wifi to stream.

I use Shield TV mainly. But usually use the nvidia official app, because it uses h.265 and moonlight uses h.264 on shield tv.

I use also an android box, mini m8s II (amlogic s905x) with moonlight and h.265.

Ethernet card on pc set to 100Mbps full duplex, because it gives me unplayable stutter if set to 1Gbps if video bitrate is set to more than 25Mbps only on amlogic s905x. With 100 Mbps i can go to 80Mbps bitrate with no problem. Really doubt that this is a problem with moonlight. 

Everything else is already described on my posts above.

Everything that i find with my testing i´ll post it here. Thanks.

**Comment by chicozeb** [comment]
After a bit of testing with the witcher 3 and mass effect andromeda, i see that the stutters are gone and the latency is much improved. Testing with the mini m8s II android box, that has a video decoding latency of 44ms with h.265. 30 Mbps. Ethernet.

In mass effect andromeda, with the first no presentation time build, it would take something like half a second between button press and the character to jump. 
With this build feels almost instant. I press and the character jumps. And i'm using an xbox one s controller through bluetooh, so more latency added. Overall, i think it has a latency of 150ms, while the previous build had something like 500ms. Of course, this are estimates, and i'm far of being a machine...

More testing is necessary, but it's looking good. 

Very good job. :)

**Comment by redlight5** [comment]
There is definitely improvement with all the other changes included in this build. Latency is now bearable and I would say good enough for slower paced games (my ultimate test game is always Binding of Isaac, that game is tough with any lag and still too tough with this build). But great improvement, thanks!

**Comment by JasonS7notG5** [comment]
Feels better than the 5.0.1 based APK. When I stream in 720p 60fps mode (game running at 1080p 60fps) with USB tethering it feels almost 100% perfect (latency is 4ms). Streaming in 1080p 60fps increases the latency to 8ms (also in USB tethering mode), this is when it gets noticable so playing racing games or fast paced shooters is not really possible. On WiFi I get 5ms on 720p 60fps and 10ms on 1080p 60fps but latency is noticable on both scenarios but this may be my setup since I only have 60% signal strength (PC also on WiFi, not via Ethernet) and my WiFi stick has only one antenna while the S7 edge has 2 antennas (dual band). Well I guess I need to reposition my router and buy a better WiFi stick and maybe connect my PC via Ethernet for the best results.

When will you implement this fix in the official build? And is there no source code anymore since v5.0.0? I ask because I would like to see this fix included in Stream Theater and VRidge which make use of Moonlight.

**Comment by cgutman** [comment]
@JasonS7notG5 I'm not sure that the version of the fix here is shippable in a real release. I don't think I would like to regress latency as badly as the fix does. I'm trying to find a way to get the same frame pacing improvements without stuffing the rendering queue and increasing latency.

Source for official builds is still being pushed to the repo here as always. I haven't been publishing APKs here, but I'll get back on that for v5.1.

**Comment by chicozeb** [comment]
@cgutman Maybe you can add the fix on moonlight settings like an selectable option. Like:"select this option to reduce stuttering at expense of latency" or something like that. I have no idea if it´s possible or if the setting must be enabled during instalation.

I´m saying this because most people probably cannot find this thread and thus cannot use this fix. While installing .apk is easy for some, i´m sure there are people that have some difficulties. And then, if they find stuttering they simply think that your software is not very good and don´t use it.

Of course, this only while you can´t fix the stuttering by other means.

But i cannot find any latency difference between this build and the official nvidia gamestream app when i run moonlight on shield tv ethernet connected. Maybe it´s there, but i cannot notice it. And i think that´s a good thing.

**Comment by JasonS7notG5** [comment]
I am hoping that you will find a better way to eliminate the micro stutters without increasing latency. Looking forward to that day. :P

**Comment by devilxxmaster** [comment]
Me pasa igual, tengo un equipo bastante potente he probado con una tablet Lenovo Yoga Tab 3 Plus con 8 procesadores y 3gb de RAM, y tengo un lag imposible de mantener. La tablet tiene Android 6.0.1 , sin embargo, tengo un Samsung Galaxy S7 Edge con Android 7.0 , y en este funciona perfectamente...
He probado bastantes cosas pero nada. Antes de encontrar este post hice este vídeo:
https://youtu.be/FUDQflHbMeM
Después encontré este post y estoy viendo que es un bug (por lo que puedo deducir por lo que leo aquí con mi pobre nivel de ingles)
Todo mi apoyo a los desarrolladores y espero que lo solucionen pronto porque la app me encanta.
Ya he probado a reinstalar los drivers nvidia con una instalación limpia, habiendo pasado el 3dguru, y he probado las otras dos versiones de la aplicación del post, pero no se soluciona. 

Siento escribir en español pero mi nivel de ingles escrito es demasiado bajo.

It happens to me, I have a pretty powerful computer I have tried with a tablet Lenovo Yoga Tab 3 Plus with 8 processors and 3gb of RAM, and I have a lag impossible to maintain. The tablet has Android 6.0.1, however, I have a Samsung Galaxy S7 Edge with Android 7.0, and in this works perfectly ...I have tried quite a few things but nothing. Before finding this post I made this video:Https://youtu.be/FUDQflHbMeMThen I found this post and I'm seeing that it is a bug (so I can deduce from what I read here with my poor level of English)All my support for the developers and I hope they solve it soon because the app I love.I have already tried to reinstall the nvidia drivers with a clean installation, having passed the 3dguru, and I have tried the other two versions of the application of the post, but it is not solved.I'm sorry to write in Spanish but my level of written English is too low.

**Comment by PhantomLiberty** [comment]
Newest fixed build doesn't change the latency at all from the first one. Tested with USB tethering. Still getting anywhere from 18-26 ms of lag at 1080p/60 but it feels like way more. Framerate is still smooth as hell which is great but the lag kinda kills it. Mainly because I can get 5ms at 1080p/60 on the other builds which feels virtually lagless.

I will be so happy if one day we can get the best of both worlds with low latency and smooth video playback.

**Comment by JasonS7notG5** [comment]
I hope the devs didn't put on hold or abandoned the work for the micro stutter fix (without increasing latency). I don't know how Moonlight exactly works but maybe it needs to be updated on par with the newest GameStream app from NVIDIA since Moonlight is a reverse engineered app from that one. And GeForce Experience got an update which has "reduced video stutter by matching monitor refresh rates of client and server". Maybe have a look at it and update Moonlight if possible please.

**Comment by PhantomLiberty** [comment]
Any progress on this matter?

As smooth as the fixed build is I still can't bring myself to use it because of the lag. Opposite for the Play Store version, can't bring myself to play with it because of the annoying stutter but the low latency is nice.

**Comment by cgutman** [comment]
Here's a v5.2.1 with the microstutter hack - https://drive.google.com/open?id=0B_5O1xNt8Q9lUEZtbFNkT3c2RWs

I'm going to just include an advanced option for it in the next build.

**Comment by WizLizard** [comment]
@cgutman I also have this issue on a standard Galaxy S7 and an Axon 7. It's literally the only thing preventing moonlight from being an enjoyable experience. The hack eliminates the microstutter/frame pacing issue but the extra input lag is unbearable. Please i'm begging the devs to make a proper fix for this one of your main priorities because it's a real shame and makes it all to good to be true.

**Comment by kbauer** [comment]
I did some more testing on this issue.

- **In Flight-Mode with enabled Wifi (i.e. mobile chip disabled) the stuttering disappears.**
- A hickup happens roughly every 2½ to 3 seconds, with both sound and video skipping some frames.
- Stuttering still appears with the "never drop frames" fix.
- Stuttering does not appear on any other device I own (iOS 10 via iPad Air 2 Wifi-only, Android 5.1 via Motorola Moto G 2013 without Sim-Card)
- Stuttering also occurs in other Streaming apps (specifically the VR app "Stream Theater").

I.e. the Galaxy S7 seems to "do" something every 3 seconds on mobile networks that interferes with streaming.

**Comment by jaysharp29** [comment]
There are a couple different things happening at the same time which conflate these issues.

Not all phones refresh at a flat 60hz.  My moto g4 uses a lower 59hz and moonlight doesn't support variable framerates.  Every one of my phones that uses a proper 60hz works with the "never drop frames" setting,  but not the g4.  It will drop a frame at 60fps nonstop. 

This issue was brought up to the developers of vridge, and they fixed it right away.  I'm not sure moonlight will ever get the ability.

Also, gamestream forces vsync on, and even the shield suffers frame skip when the host and client don't stream at the exact refresh.  Gamestream doesn't offer a headless mode, which would probably help.

Finally, almost all lollipop os phones work flawlessly.   I downgraded a phone from marshmallow to lolipop and all problems went away.  Not a skip in sight.  Too bad newer hardware doesn't support lollipop.

**Comment by kbauer** [comment]
@jaysharp29 In checked by using www.displayhz.com.

iPad Air 2 and Moto G, where streaming works regardless of mobile network connection, have a refresh rate of 60Hz straight.

For the Galaxy S7 an odd refresh rate of 59.67Hz is reported, which would correspond to a mismatch of one frame per 3 seconds (and 3 seconds is the observed stutter interval). No idea though how this fits with the issue vanishing upon using flight mode.

**Comment by jaysharp29** [comment]
I don't know how flight mode affects the hardware.  But the odd refresh is fixable by the developers.  Moonlight doesn't have to request 60fps. Trinus, vridge, even kinoconsole allow customizable fps.

**Comment by saiotek** [comment]
I have the same problems with my z3 tablet. Downgraded to 5.1.1 and the microstuttering is gone, but some other framedrop is still there.
Displayhz.com says 59.39hz
So i changed 60 to 59 in the config, build the app and set the game to 59 on the host pc and now it's hard to find framedrops. Decoding numbers changed from 30ms to under 20ms

The problem i see with customisable fps, you have to set it in moonlight and on the host pc...

**Comment by jaysharp29** [comment]
Can you upload that build? Or explain how to change it? I haven't tried building moonlight.

**Comment by jaysharp29** [comment]
Can a 59hz build be uploaded?

**Comment by saiotek** [comment]
I can upload the build, have to get some freespace first. i deinstalled the original moonlight. I use the msi afterburner with rivatunner to set the fps to 59. There is although an option for 50fps.

**Comment by saiotek** [comment]
https://mega.nz/#!hmwEWboS!ov4twzICDX6cm52FQVRjurfkj1BGARq-Dye5JdmZRLA

Just for testing ;)

**Comment by jaysharp29** [comment]
I'm trying to create one myself but can't figure out what in 'preferences'
actually controls fps. Learning github and Android studio has been eye
opening.

**Comment by saiotek** [comment]
config.fps see pull request #123

**Comment by jaysharp29** [comment]
Almost perfect with the never drop frames settings.

Wonder what else could be done?

**Comment by jaysharp29** [comment]
I figured it out!! No latency. Create a custom resolution in windows with a
58.994 refresh.  Bypasses the lag on this build.  It's as fast as regular.

Must skip some type of vsync.

**Comment by jaysharp29** [comment]
I get 16ms at 1080p 59 hz, no stutter.

A 720p 59 would be faster still.  I'll add it myself
 Thx for the app.

**Comment by jaysharp29** [comment]
Last update. Using rtss with a custom 59.993 limit brought the latency down to 13ms over wifi.

So I hope this option is somehow pushed to main, because new gfe will probably stop this from working in the long run.

**Comment by jaysharp29** [comment]
I meant 58.993.

**Comment by saiotek** [comment]
It should be easy to just make an option under the microstutter fix like manual fps override where you can choose none (default), 50fps, 59fps and this overrides the config.fps for all resolutions...

**Comment by WizLizard** [comment]
This won't help for emulation. Games running through emulation need to maintain their target framrate in order to achieve full game speed, i wouldn't settle with this kind of fix.

**Comment by jaysharp29** [comment]
That would excellent, although like you said, it requires instructions on
the host machine.  My diy vr experience is pretty much flawless now.

@cgutman, any chance?

**Comment by jaysharp29** [comment]
The streaming quality itself vastly improves just picking the correct refresh rate.  Unless android has the ability to correct  or use adaptive sync, how else would solve this?  

But yes, not everything likes an odd refresh.

**Comment by WizLizard** [comment]
So what happens if the refresh rate remains 60 on the host machine and the fps is limited to 59 for the stream, does stutter still occur?

**Comment by jaysharp29** [comment]
Haven't tried it.  Any emulator in particular you're curious about? I have
dolphin.

**Comment by WizLizard** [comment]
A few actually. The ones that aren't capable of running on a phone with raw emulation. Like PCSX2, CEMU, and yeah Dolphin even though it's getting there.

**Comment by jaysharp29** [comment]
I'll try dolphin and pcsx2.

**Comment by jaysharp29** [comment]
Doesn't work in dolphin.  It's the reverse of regular moonlight now, 60 in,
59 out, a skip every frame. Some games have variable fps, but this isn't an
emu fix, at least for dolphin.

**Comment by jaysharp29** [comment]
Are you sure your phone even uses lower 59hz? This is only a fix in that
situation, unless you prefer 50fps.

**Comment by WizLizard** [comment]
It's like 59.67 or something.

**Comment by jaysharp29** [comment]
Perhaps the phone can use all refresh  below 60 and it will work.  This
only works with "never drop frames".

**Comment by jaysharp29** [comment]
To get the lowest latency, I have to switch to 59hz refresh on host pc
after connecting.  So you should sync at 60hz and then pick a 59hz
resolution.

Starting the host pc with 59hz gives me 20ms.

On Jun 6, 2017 10:46 PM, "redlight5" <notifications@github.com> wrote:

Made a small video to demostrate. First half is the official build which
has stuttering (visible in the video) but great low latency. Second half is
the unofficial build linked here without stuttering but bad latency. The
reported latencies in this test were: Official 13ms, Unofficial 17ms (it's
definitely more).

https://drive.google.com/open?id=0B4FaEFFulZWINWZsRW9tYWNaZzA

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/moonlight-stream/moonlight-android/issues/349#issuecomment-306678430>,
or mute the thread

**Comment by WizLizard** [comment]
Not all games detect a 59hz signal even steam games. For example Star Ocean The Last Hope and Cuphead.

**Comment by kbauer** [comment]
Additionally, the frame rate mismatch issue is probably separate from the Galaxy S7 [Edge] issue, given that for the S7 stuttering vanishes in Flight mode.

By the way, can another S7 / S7 Edge owner confirm this behaviour? To reproduce: Enable flight mode, then reenable ONLY wifi.

**Comment by WizLizard** [comment]
I'll test It

**Comment by jaysharp29** [comment]
Flight mode disables the radio, which leads to power savings.  Maybe the s7  uses more aggressive display sync when the data services are disabled.

As long as the phone actually uses the upper range of 59-60hz, you're probably right about it being a different issue.

**Comment by jaysharp29** [comment]
Yeah, not all media will work at 59hz.

But the alternative for a phone like mine is skipping in everything.  I'll take the tradeoff.

**Comment by kbauer** [comment]
Apparently, setting the **screen-resolution** to **HD = 1280x720** removes the stuttering for the Galaxy S7.

This can be achieved by using the "Performance Mode" setting in the notification-bar menu, by customizing the settings of one of the modes accordingly. It doesn't matter if one uses "Game Mode" or "Performance Mode" for this.

![screenshot_20180105-011814: Press and hold "Performance Mode" button. Press "Game". Press "Customize". Select "HD" screen resolution, then press "Apply".](https://user-images.githubusercontent.com/7498468/34590717-d0617a1a-f1b8-11e7-8806-d3119e796ef2.png)


I switched a couple of times, and tried several "resolution and framerate"-targets. Apparently nothing other than selecting "HD" helps. E.g. setting the resolution to Full HD and the resolution target to "1080@60Hz", produces the same stuttering, as setting to resolution to WQHD (native resolution) with "720@30Hz".

In some cases, I had to restart the game after changing the resolution, as otherwise the software would reject the connection.

**Comment by jaysharp29** [comment]
Did you ever try the refresh fix?  720p on an s7 is a big compromise.

**Comment by kbauer** [comment]
Probably relevant: I had "Never drop frames" disabled during my recent tests (Flight Mode, HD resolution).

@jaysharp29 

**720p.** Not much of a compromise on the S7, since Samsung bundles software to quickly switch these settings right into the notification bar. Worst-case I'll forget to switch off game-mode, and will have to disable it (takes a few seconds) once I notice the blurryness. Once set up, it is a single press of the notification-bar button.

**Refresh fix.** I am currently trying to build moonlight myself, but I'm not sure where I have to change this. My best guess would be the lines `config.fps = ...` in `./moonlight-android\app\src\main\java\com\limelight\preferences\PreferenceConfiguration.java`, but I would have expected a floating-point number here, not an integer.

**Comment by WizLizard** [comment]
@kbauer So with flight mode on and never drop frames disabled you can leave the phone resolution above 720p with no microstutter?

**Comment by kbauer** [comment]
@WizLizard Yes, with Flight-mode there is no stuttering even at native WQHD resolution.

@jaysharp29, **refresh fix**. I tried the build posted by @saiotek above, and it didn't change anything. Neither did it enable configurations that didn't work before, nor did it sabotage those that worked.

**Comment by jaysharp29** [comment]
Intel label 59.94hz as 59hz.  It's not 59.000 as you would need for the refresh build.  The refresh would need to be changed in nvidia custom resolutions, which you may not have access to with optimus.    

But your fix is probably the proper one for s7 once other people verify it.  

Finally, the refresh fix gives me almost the same latency (12 vs 9ms) as moonlight without ' Don't drop frames', which could be seen as a best of both worlds.

**Comment by WizLizard** [comment]
@kbauer Tested the airplane mode solution and the custom 720p solution, neither worked on my S7 there were clearly still frame pacing/microstutter issues.

**Comment by saiotek** [comment]
Which version do you have? Normal or Edge? Exynos or Snapdragon?

**Comment by WizLizard** [comment]
Normal, Snapdragon.

**Comment by kbauer** [comment]
@WizLizard For clarification: 

  1. Do you have the same version of Android? I have 

         Android version               7.0
         Kernel version                3.18.14-12365438
         Android security patch level  1 December 2017
        
  2. How does the stuttering look for you? The major issue that the
     two methods solved for me, was a very noticable hickup
     periodically every 3 seconds, while some minor stuttering
     (smaller duration, not periodic, sometimes not at all) remains on some devices depending on the situation (wifi-tablet distance, use of Bluetooth peripherals, etc).

**Comment by WizLizard** [comment]
My kernel version is 3.18.31. It is not stutter generated from a connection issue. It's inconsistent frame pacing. I use an xbox one controller via Bluetooth. I've tested other controllers but don't think it's related. It's easily visible in 2D side scrollers/platformers. There's no audio or visual artifacting as it's happening.

**Comment by saiotek** [comment]
checked some devices:
S6 59,72hz
S5 60,41hz
S4 60,17hz
Note4 59,90hz

**Comment by jaysharp29** [comment]
Is the snapdragon s7 excluded from the airplane mode fix then?  Exynos may handle frame delivery differently.

**Comment by PhantomLiberty** [comment]
I would assume so. I tried the airplane mode method but the stutter still remained. It was not more or less stutter but the same amount.

**Comment by jaysharp29** [comment]
If you have stutter:

1.Download the 59hz build.  
2. Create a refresh of 58.94 in nvidia custom resolutions.
3. Set moonlight to never drop frames. 
4. Play a game in windowed mode to verify the 59hz resolution is used. 
5. Turn on vsync in game.
6. Use a framelimiter like dxtory or redd to get the best results.

It shouldn't skip at all.  If it does, go to www.displyhz.com and see what refresh the phone uses.  I think 59.5 and down should work fine at 59hz, but I'm curious if all phones can accept a 59hz refresh, like pc monitors.

**Comment by JasonS7notG5** [comment]
Thanks to @saiotek and @jaysharp29 the microstuttering is gone!
I just tested CoD:MW, GTA V and BF4 Multiplayer. Everything is so much better now, this is the proper fix for microstutters without increasing latency.

I just did what jaysharp29 said, I downloaded the 59FPS build from saiotek, created a custom resolution in the NVIDIA Control Panel (1080p 59Hz). I downloaded Rivatuner Statistics Server and set the framerate limiter to 59. Then I chose the 59FPS and "never drop frames" option within Moonlight. And it worked. So easy. I could even play multiplayer in BF4 with only little disadvantages.

Using the offline mode and 720p resolution on my S7 didnt made any difference, I tested that months ago. My screen has a refresh rate of 60.30 Hz.

I dont know if it made that much of a difference when I tested games with fullscreen with vsync, fullscreen without vsync, borderless windowed with vsync and borderless windowed without vsync. I prefered borderless windows without vsync but maybe thats just a placebo. Vsync for me looked like it was not needed since like one of you said GameStream has vsync enabled by default, therefore I couldnt see any screen tearing.

Now implementing this fix in the main build would be nice and also for 720p since that has even less frame decoding latency.

**Comment by jaysharp29** [comment]
@JasonS7notG5 Thanks for confirming that 59hz is an acceptable workaround even on 60hz phones.  

I recommend borderless and window mode because that is the only way to be sure your custom resolution is used.  But if you are positive it will be used in fullscreen,  it should be even faster (no aero being forced on).  There are also a few things that don't like 59hz, like emulators and some pc games.

Make sure you sync at 60hz and then pick a 59hz resolution after connecting.  That dropped my latency all the way down from 20 to 12ms.  


I guess if this doesnt make it into main, we'd have to create a new branch that's always up to date.  Doesn't seem too bad.

**Comment by kbauer** [comment]
Sadly, despite no apparent changes to the setup, the screen-resolution method described above has stopped working for me on the Galaxy S7. Flight-mode still works.

**Comment by jaysharp29** [comment]
What is the charge level of the phone when it does work?  It could be a power savings thing.

**Comment by kbauer** [comment]
@jaysharp29 Tested instantly after recharging today. Yesterday the battery indicator was at around 70%. Shouldn't be a battery thing therefore.

**Comment by cgutman** [comment]
Running "60 FPS" at 59 Hz is something I had considered (and even tried).

The core latency issue here is the server getting slightly ahead of the client's decoding and causing the client's buffer queue to fill faster than the screen can empty it. That's the whole reason for the frame drop logic - it prevents the decoder's output buffer queue from generating a huge frame backlog for the display. The problem which makes this difficult on Android is that it's seemingly impossible to introspect the display buffer queue to determine whether it's running behind. I have to infer the state of the queue based on other callbacks.

In Moonlight, there are 2 queues in play on the output path:
[Video Decoder] -- MediaCodec Output Buffer Queue --> [Moonlight] -- SurfaceFlinger Display Buffer Queue ---> [Display]

In Lollipop and earlier, Moonlight could only drop frames when processing the first queue (we could tell the codec not to send any particular frame to the display). However, Marshmallow started also dropping frames in the second queue if they had presentation times that would put them on the display at the same V-sync window. This additional drop lowered latency a bit, but also caused tons of frames to be dropped if a small bit of network jitter or client/server time sync issues caused us to receive 2 frames in the same V-sync window. The "Never drop frames" options disables dropping at the second queue by not submitting any presentation times on frames.

The easiest way to generate enough negative pressure on the display's buffer queue is to have the PC encoding slightly fewer frames than the screen can display per second. This ensures if network jitter causes a set of frames to be delivered at the same time, we will still empty the display's buffer queue in short order to get back to low frame latency, otherwise the lost time will increase the stream's latency forever.

The issue with just going with 59 Hz is that it guarantees that we drop at least 1 frame per second, which sucks (though probably less than what's going on now). A better solution would be to dispatch frames to rendering from Choreographer where we can be sure the display just consumed a frame before adding a new one.

A couple GFE releases ago, Nvidia also added a more precise client refresh rate option to reduce stuttering caused by mismatched refresh rates on both devices. Our current FPS option only accepts integer values, but this other option accepts fractional frame rates. However, I did try messing with this option and couldn't seem to be able to get it to throttle frame delivery to match my refresh rate, even when I specified something where throttling would be obvious like 30 Hz (maybe it only works with certain ranges).

Hopefully, combining Choreographer and the precise client frame rate option will yield something workable.

**Comment by JasonS7notG5** [comment]
So the elegant variant would be the more complicated solution to implement so we wouldnt need to limit our FPS to 59 or lower on the Host PC. Thats what I understood at least.

But can we in the mean time have a "Limit FPS to 59" option in Moonlight under "Never drop frames" with a note that the Host PC also needs to limit the FPS to 59? Thats what we have right now with the custom build from @saiotek but implementing it to the main build would be nice. Thanks.

**Comment by jaysharp29** [comment]
@cgutman Aren't we actually requesting 59 fps from both the device and the host?  You say this skips a frame, but on my phone, the actual vsync is only 59hz.  That's why the custom resolution and fps was created.

The 59hz build seems to run perfectly on my phone without frame drops.  Even with the "never drop" setting I'm getting 12 ms latency.  This is after creating a 59hz resolution on the host.

For my particular device (moto g4), I only need access to the same setting on the newer master.  Saiteks build will probably be obsoleted with newer GeForce experience updates, which is a concern.

This is a great workaround until you accomplish what you feel is the proper fix.

**Comment by jaysharp29** [comment]
New gfe of course broke the 59hz build.

**Comment by JasonS7notG5** [comment]
The 59FPS build here is for GFE 3.11. Moonlight got updated for GFE 3.12 but we dont have the 59FPS option inside the main build. @cgutman please just add a 59FPS option like @saiotek did, I think that shouldnt be hard or time consuming. I understand you want a better solution than this but this solution right now works, I dont see any frame skipping, it feels just like how it should. Like I said, I was able to play a MULTIPLAYER game with very little disadvantages (and thats only because of the decoding latency and my PC is connected via 5GHz 1x1 WiFi stick to my router with a WiFi stregth of 60-80%).

**Comment by jaysharp29** [comment]
@JasonS7notG5 I downgraded to gfe 3.11 and used the window host file to block the update.  I can post instructions or you can find them on the internet.

Unfortunately, this is the only way I could continue to use the 59hz build.

**Comment by JasonS7notG5** [comment]
I am waiting for @cgutman to integrate this to the main build, I dont want to downgrade and block updates etc. 

@cgutman This fix works! It really works! It doesnt skip anything. Yeah I am not an programmer, you may "see" stuff under the hood which I cant, you may think this is not a good idea to implement but it is. You know I played various titles thousands of hours and I know if a game has micro stutters, has latency etc. And like I said, I could not only play multiplayer matches in BF4 (which is a fast paced shooter), I also enjoyed it. I have not noticed any frame skipping, this is the deal, this is the solution. Give us, the ones who know a bit more than the average user, the option to limit Moonlight to 59FPS. The majority of users dont know or want to make a custom resolution with 59Hz on their Host PC and download and install third party software like RTSS to limit the FPS to 59. But we know how and we want this option please.

**Comment by JasonS7notG5** [comment]
Sorry but I dont agree. I am also "working" by testing and reporting issues. Others and I have brought the micro stutter problem to light with this thread here. I spend many hours of my time showing the problem Moonlight had / still has by making videos, providing logs etc. Why is there an issues section where people can report whats wrong and at least in this case also give a solution how to solve it? Why should this be a separate build from an already not official solution to GameStream? It is confirmed that everybody on Android 6.0 has micro stutters, every single one, at least on smartphones via WiFi (no one will use an Ethernet adapter for smartphones). Why hold off this solution to the thousands of people using just the main build from the Play Store? This is a simple fix from @saiotek, he even told how to do it: "just changing the config.fps in PreferenceConfiguration.java and creating a new resolution with that fps on the host pc would do". You know what? Because we wrote so much here @cgutman chose to implement the "never drop frames" option inside Moonlight. The community helps making Moonlight better, not making hundreds of different builds and separating features etc.

**Comment by jaysharp29** [comment]
Fair enough.  I definitely want it fixed in main because GeForce experience will continue to evolve.

**Comment by JasonS7notG5** [comment]
Tested it with the Play Store main build, 1080p 60fps option on Moonlight with never drop frames and on Host PC custom resolution 1080p 59Hz and FPS limit to 59 with RTSS. It is better than standard 60 Hz and 60+FPS but the 59FPS option in Moonlight was better, it really eliminated the problem completely. 

So @cgutman would you please respond to us now and tell us what you are going to do or not and why.

**Comment by cgutman** [comment]
Here is a new 59 FPS variant APK: https://drive.google.com/open?id=14pR5C-ChigPvwO1wXS1pxkr6LVX1ulQA

And the change in that build: https://github.com/moonlight-stream/moonlight-android/commit/c87058bd9505493a86866ee831aadda605a76624

**Comment by jaysharp29** [comment]
Thank you.  Will test it out to see if it's identical to the saitek fix.

**Comment by jaysharp29** [comment]
This version is perfect for my moto g4.  At 720p 59 fps, never drop frames, I get no stutter and 6ms over wifi.  This is on gfe 3.11 since I blocked the update.

Thanks.

**Comment by cgutman** [comment]
I am experimenting with new rendering code to smooth out frame delivery. Can you guys please try this build and enable "smooth mode" (what used to be called "never drop frames") and tell me how it goes? https://drive.google.com/open?id=1r4VwDXxLNaxbFWOIYbt9VGffMNY-YlJC

**Comment by cgutman** [comment]
It looks like the frame delivery on that build is faster than "never drop frames" but still worse than the 59 FPS hack. Alas, I think for now I will release the hack, since it undoubtedly improves the situation for users on M and higher.

The default will be to stream at one FPS less than the display with no frame drop (current behavior with the 59 FPS hack build). The "never drop frames" checkbox will now force regular 60 FPS no-frame-drop behavior.

**Comment by JasonS7notG5** [comment]
I made some quick tests with BF4 on the new Moonlight version from the Play Store. I have less latency (both average frame decoding and hardware decoder) now compared to the previous versions. I reset my NVIDIA Control Panel resolution back to 1080p 60Hz and closed RTSS, so the framerate is not locked at 59. I then played some minutes on default without the "never drop frames" option and some minutes with it. I cant really tell the difference between them. It feels like the default setting is perfect with vsync off but having vsync on feels less responsive. When using the "never drop frames" option I cant really notice any difference. In this, vsync on or off feels great, just like when on default + vsync off. I havent noticed micro stutters on any of them. But thats highly subjective right now, could be a placebo effect. 

I also tested the old trick settings: custom resolution to 59Hz and limiting the FPS to 59 with RTSS on default without the never drop frames option, feels great but I think it is not needed. Having Moonlight on default works just as fine. Need to test BF4 further on the test range and multiplayer and also other games like GTA V by driving a fast car and dodging oncoming traffic.

**Comment by jaysharp29** [comment]
So the new version requests one less frame than the host?  Do you need 60hz refresh for 59 fps then?

**Comment by JasonS7notG5** [comment]
Bring everything back to default, so your NVIDIA Control Panel resolution back to normal (probably 60Hz), close RTSS FPS limiter and restart your PC. Use the default Moonlight settings (1080p 20Mbit/s) and dont make a tick on "never drop frames". Play your game with vsync off ingame. Write your impression here afterwards.

**Comment by JasonS7notG5** [comment]
After some more testing with BF4 multiplayer and GTA V, it didnt really matter if default with vsync on/off or "never drop frames" with vsync on/off. Felt all the same to me. Feels good. I get better latency when using 720p, it is noticeable slightly better than 1080p when playing fast paced multiplayer games. Good job, micro stutters are gone.

But what exactly is the difference between the default and "never drop frames" option besides one running with 59 FPS inside Moonlight and the other with 60 FPS? What have you changed, because before having the 59 FPS hack without "never drop frames" ticked still had micro stutters. Well I dont know but the default or "never drop frames" option is obsolete, thats how I see it at least on my testings. Or it needs to be called differently but even then who is gonna supposed to play on default and who with "never drop frames" ticked?

And is it possible to deactivate the vsync applied by GameStream? Because I think turning that off would reduce the latency / input lag even further.

**Comment by jaysharp29** [comment]
After testing the newest version, I find that I still have to set 59hz in the control panel to avoid stutter.  I believe this is because my phone runs a 59.2 refresh as opposed to 60hz.  So this is not technically a universal fix because some people (like myself) will still have to go into the control panel to change refresh.  But once that is done, it is identical to the other build in frame delivery. 

But latency is worse.  On the 59 hz build I can get latency down to 13 ms by switching resolutions mid stream.  That trick didn't work for me here.  So I'm hovering at 20ms.  

I think the new method could have the ability to confuse people who already had a decent experience.  I wish moonlight didn't have to set a fps cap at all.  Just enable vsync and use whatever fps the host provides, while not dropping frames.

**Comment by cgutman** [comment]
@jaysharp29 how are we supposed to "use whatever fps the host provides, while not dropping frames" if the host PC outputs more frames per second than the client can consume? That is exactly the issue here - 60Hz host with 59 Hz client.

V-sync on or off is not a question. The host PC will capture however many frames the GPU outputs up to the FPS limit specified by Moonlight, whether v-sync is on off. Android imposes v-sync with no way to opt-out.

Even a host outputting 60.1 and a client consuming 60.0 leads to frame latency runaway without dropping frames.

Moonlight will set the FPS limit to your display refresh rate (as reported via Android APIs) rounded to the nearest whole number minus 1. You can use logcat to see what FPS limit Moonlight has chosen for your hardware.

**Comment by jaysharp29** [comment]
Further testing shows that you can leave the host on 60hz and use a frame limiter in place of a refresh change.  But vsync has to be on.  So, I leave the new moonlight at default, set 60hz on the host pc, but use rtsss to cap the game to 59 hz.  I get the same the same result as using 59hz refresh.  

@cgutman, I probably didn't explain what I mean.

The example I give only pertains to my phone.  In kinoconsole and vridge, two other streamers, If I enable streaming, and use a framelimiter, I only get the fps that the limiter provides.  So vsync is set at 59 hz, and I get 59 frames delivered to the phone.  Only moonlight didn't work like that.  Not until you explicitly requested 59fps.  

So if vsync is set, and it caps at the host refresh, do you still need to request a number of frames?

For instance, your new version works like those other programs now.  I can set 60hz, but use a 59 fps cap  via another program and still get no stutter.  So the 60fps number seems arbitrary.

**Comment by cgutman** [comment]
@jaysharp29 maybe I'm still not understanding the question. Moonlight has always imposed a cap on frames captured per second from the host - that's how 30 and 60 FPS settings work. Even if the PC is rendering 144 FPS, the server will only capture 60 per second. This is also how the 59 FPS option works, it limits capture to 59 FPS even if the PC is rendering more. The capture rate may drop below the limit too if the PC is rendering fewer frames per second than the cap.

You still need to request an FPS cap with V-sync because the host and client refresh rates may not be the same. The cap is there to ensure the host is not encoding more frames than the client can decode and display in real-time.

**Comment by jaysharp29** [comment]
Thanks for the explanation.

**Comment by JasonS7notG5** [comment]
https://www.testufo.com/refreshrate

So my TV shows perfect 60.000 Hz and my S7 edge shows around 60.300 Hz. So having Moonlight on default will round the display refresh from 60.300 Hz to 60 Hz and then subtract with 1, so in the end I have 59 Hz on the S7 edge? And the "never drop frames" option gives me 60 Hz?

My TV (Host PC) also shows me perfect 60 FPS while on the S7 edge I get "stutter warning" and "performance warning" with the Samsung Internet and Chrome app. I can see the FPS counter dropping sometimes to 50 to 60 range (this was the same problem with prior Moonlight versions back when I tested it with GameBench). I dont know how Moonlight handles it now but it feels good and I dont notice micro stutters anymore.

**Comment by TehJim08** [comment]
@Trasen Can confirm that deleting old drivers and reinstalling the new Nvidia ones fixed this problem completely. was originally getting constant stuttering every 5 seconds on 1mbps bitrate on 720p 60fps. I can now play at 1080p 60fps with over 20mbps bitrate!

**Comment by PhantomLiberty** [comment]
Anyone else getting increased latency lately? After the stutter fix at 720p60fps I would get 5-6ms but now it's around 12-16ms. It's still worse but never skip frames but it's definitely worse than it used to be. Pretty much back to the old days of having a smooth stream but being very laggy. I would really like to start using Moonlight again but with this sudden increase in latency I don't see that happening as it's pretty unplayable to me compared to what it used to be.
